### PR TITLE
Add CSpell

### LINKED ISSUE #681: Add Cspell
# Problem

Without a spell checker , spelling mistakes can go unnoticed. However, there are some words which are domain specific.

# Solution

Add Cspell and a cspell.json file to whitelist certain words

# Alternatives considered

# Additional context

### PR DESCRIPTION
**Pull request by kevaundray**
# Related issue(s)

Resolves #681 

# Description

## Test additions / changes

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

# Additional context

### CODE REVIEW

**Comment by kevaundray** [comment]
Note that the json file also includes words which are misspelt -- This is intentional and once they have been removed and corrected, the PR will be ready for review

**Comment by kevaundray** [comment]
To print out all of the unknown words, one can install the cli and run:

```
cspell-cli "crates/**/*.{md,rs}" --words-only --no-summary --no-progress --unique
```

**Comment by kevaundray** [comment]
A large group of the spelling mistakes fall into two categories:

- Mispelled words (Seems to be about half of the list)
- incorrect casing; for example `myfunc` is seen as one word, but `my_func` or `myFunc` will be recognised as two words


Then we have domain specific words like acvm, acir and rustc which should indeed be added to the list.

**Comment by kevaundray** [comment]
In the worse case, if we keep adding new words to the codebase one will need to keep updating the list.

The upside is that we no longer need to manually check for spellling mistakes from each contributor and the introduction of a new word will immediately be obvious (warranting an explanation)

File: cspell.json [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        "puposes",
+        "flattenning",
+        "adress",
+        "lenghts",
+        "conresponding",
+        "initialised",
+        "occured",
+        "optimised",
+        "initialise",
+        "contrain",
```
Comment (jfecher) [inline cspell.json:25]: Is this the list of allowed misspellings? I don't think we should allow "contrain" or "concatened" for example.

**Review by jfecher — COMMENTED** [review]
Does this PR fix any issues with the codebase? I'm somewhat dubious of the need for a spellchecker. I don't see it solving any issues and it just adds an extra CI step.

**Comment by kevaundray** [comment]
> Does this PR fix any issues with the codebase? I'm somewhat dubious of the need for a spellchecker. I don't see it solving any issues and it just adds an extra CI step.

We do have quite a lot of spelling mistakes -- if you check the cspell.json file, I have added the words which were misspelled into the json file to indicate them and to have the CI pass. 

The top ones are spelling mistakes and in other places, it is because we are not using the snake case to correctly separate two words for example.

Getting it in the CI is a pain -- we can also opt to make them warnings in the PR instead of errors, though I don't expect us to add a bunch of new unknown words, so I'm up for trying it out and seeing how often it catches new spelling mistakes introduced by a PR

**Comment by jfecher** [comment]
If it is not too much trouble then, I think this PR should also fix the spelling mistakes it finds rather than committing an ignore file with many mistakes we do not want ignored. If it is too much work, you can pass it off to someone else on the team.

**Comment by kevaundray** [comment]
> If it is not too much trouble then, I think this PR should also fix the spelling mistakes it finds rather than committing an ignore file with many mistakes we do not want ignored. If it is too much work, you can pass it off to someone else on the team.

I agree -- to clarify, this PR is not ready while we are ignoring spelling mistakes. This was just done so that the CI can pass and we are aware of the list of problematic words.

I then removed a word from the list to check that the CI failed and now it's a matter of correcting the spelling mistakes present in the code and removing them from the Json file. I will do this and then re-request your review :)

**Review by jfecher — DISMISSED** [review]
👍

**Review by jfecher — APPROVED** [review]
👍

**Pull request merged**